### ISSUE TITLE
Can Lafite do the image-to-image generation?

### ISSUE THREAD

**Issue by sbkim052** [drboog/Lafite#9]
Can Lafite do the image-to-image generation?
When I fed image feature to the Generator, the generated image was not as good as the generated images from text. 
Is there another way to generate images from image input?
The image feature was extracted from CLIP as below.

```
image = preprocess(Image.open(img_path)).unsqueeze(0).to(device)
img_fts = clip_model.encode_image(image)
img_fts = img_fts/img_fts.norm(dim=-1,keepdim=True)

z = torch.randn((num_images_to_generate, 512)).to(device)
c = torch.randn((num_images_to_generate, 1)).to(device)  # label is actually not used
img, _ = generator.generate(z=z, c=c, fts=img_fts)
```

Btw, thanks for sharing this wonderful work.

**Comment by drboog** [comment]
Yes, it can do image-to-image generation, but you need to perturb the image feature before feeding it into the generator.

Basically, you can add the following after extracting the image features, before feeding them to the generator:
```
random_noise = torch.randn(img_fts.shape).to(img_fts.device)
random_noise = random_noise/random_noise.norm(dim=-1, keepdim=True)
        
img_fts = img_fts*0.25 + random_noise*(1-0.25) # 0.25 is a hyper-parameter     
img_fts = img_fts/img_fts.norm(dim=-1, keepdim=True)
```

**Comment by sbkim052** [comment]
Thank you for the reply. 
I added the random noise, but still it does not give informative images.

I have two more questions.
Don't we need the perturbation as below?

![image](https://user-images.githubusercontent.com/59387731/161360841-97e3a461-3611-4bb9-88b9-9e0c6ff88368.png)

And is it correct to feed the image feature directly to generator?
```
img, _ = generator.generate(z=z, c=c, fts=img_fts)
```

**Comment by drboog** [comment]
Hi, I can do the image-to-image generations on my end, just like examples shown in the "No text" column in Figure 6, 12, 13, 14, 15 of our paper. I think there are few things maybe you can double check:

1. The CLIP model version used in image-to-image generation is the same as used in training the pre-trained models (e.g. if the model is trained with ViT-B/32, then only ViT-B/32 can be used here);

2. I suggest using model trained with language-free method ([Model trained on MS-COCO, Language-free (Lafite-G), CLIP-ViT/B-32](https://drive.google.com/file/d/1eNkuZyleGJ3A3WXTCIGYXaPwJ6NH9LRA/view?usp=sharing)) instead of model trained with ground-truth pairs. 

3. The input images should be similar to the samples from training dataset (i.e. you can't feed feature of a human face image to a model pre-trained on MS-COCO). My suggestion is first using samples from training dataset,  if it works, then you can try some other images from the internet;

4. The image features are perturbed correctly (extract image feature -> normalize the image feature -> perturb it with a normalized noise  -> normalize the perturbed image feature);


As for your questions:

1. We proposed two different perturbations: based on Gaussian noises and trainable networks respectively. 
What you showed is the second one, while what I recommended is the first one because it is more general.

2. I do NOT recommend directly feeding image feature into the generator, we should always construct a fake text feature using the image feature, and feed the fake text feature into the generator.

**Comment by drboog** [comment]
Here is the reason why I suggest using model trained with language-free method instead of model trained with ground-truth pairs: 

Although we are trying to construct fake text features based on image features using perturbation, we actually have a very high probability to obtain a fake text feature that is out of the distribution of ground-truth text features. 

Consequentially, the model trained with ground-truth pairs do not know how to handle this out-of-distribution feature, it is likely to fail.  

On the other hand, the language-free model is trained with fake text features (perturbed image features), thus it can handle the fake text feature better.

**Comment by sbkim052** [comment]
@drboog Thanks for the reply.
I tried on what you said, and I cannot get an informative image.
As I feed an image below,
![image](https://user-images.githubusercontent.com/59387731/161388830-5a388d89-4b8c-4a70-885e-62e41d0cdb16.png)
the result I got is the one below.
![image](https://user-images.githubusercontent.com/59387731/161388840-85fdd439-3800-4306-8b06-f22e02aa69bc.png)

The image is from the coco2014 training dataset, and I got similar results on several trials.
I attached the whole code and it would be thankful if you could catch some errors in my code.

```
class Generator:
    def __init__(self, device, path):
        self.name = 'generator'
        self.model = self.load_model(device, path)
        self.device = device
        self.force_32 = False

    def load_model(self, device, path):
        with dnnlib.util.open_url(path) as f:
            network = legacy.load_network_pkl(f)
            self.G_ema = network['G_ema'].to(device)
            print(self.G_ema)
            self.D = network['D'].to(device)
            #                 self.G = network['G'].to(device)
            return self.G_ema

    def generate(self, z, c, fts, noise_mode='const', return_styles=True):
        return self.model(z, c, fts=fts, noise_mode=noise_mode, return_styles=return_styles, force_fp32=self.force_32)

    def generate_from_style(self, style, noise_mode='const'):
        ws = torch.randn(1, self.model.num_ws, 512)
        return self.model.synthesis(ws, fts=None, styles=style, noise_mode=noise_mode, force_fp32=self.force_32)

    def tensor_to_img(self, tensor):
        img = torch.clamp((tensor + 1.) * 127.5, 0., 255.)
        img_list = img.permute(0, 2, 3, 1)
        img_list = [img for img in img_list]
        return Image.fromarray(torch.cat(img_list, dim=-2).detach().cpu().numpy().astype(np.uint8))


text_input = 'The red train is on the green grass field.' #@param {type: "string"}
# img_path = '/local_data/sung/audio2scene/dataset/a2s2s/vegas_log/datas/frame_test/3000276.png'
img_path="/data/coco_moon/train2014/COCO_train2014_000000144519.jpg"

num_images_to_generate = 1 #@param {type: "number"}

with torch.no_grad():
    device = 'cuda:0'  # please use GPU, do not use CPU
    path = './checkpoints/COCO2014_Language-free_Gaussian.pkl'  # pre-trained model
    generator = Generator(device=device, path=path)
    clip_model, preprocess = clip.load("ViT-B/32", device=device)
    clip_model = clip_model.eval()

    # tokenized_text = clip.tokenize([text_input] * num_images_to_generate).to(device)
    # txt_fts = clip_model.encode_text(tokenized_text)
    # txt_fts = txt_fts / txt_fts.norm(dim=-1, keepdim=True)

    # putting image input for the generator
    image = preprocess(Image.open(img_path)).unsqueeze(0).to(device)
    img_fts = clip_model.encode_image(image)
    img_fts = img_fts / img_fts.norm(dim=-1, keepdim=True)
    random_noise = torch.randn(img_fts.shape).to(img_fts.device)
    random_noise = random_noise/random_noise.norm(dim=-1,keepdim=True)
    hyper=0.75
    img_fts = img_fts*hyper+random_noise*(1-hyper)
    img_fts = img_fts/img_fts.norm(dim=-1,keepdim=True)

    z = torch.randn((num_images_to_generate, 512)).to(device)
    c = torch.randn((num_images_to_generate, 1)).to(device)  # label is actually not used
    img, _ = generator.generate(z=z, c=c, fts=img_fts)
    to_show_img = generator.tensor_to_img(img)
    to_show_img.save('./generated.jpg')
```

**Comment by drboog** [comment]
Here are what I got with the given sheep image (reference images at the first row, generated images at the second row).

<img src="https://user-images.githubusercontent.com/47047889/161399066-e337719d-e6b9-4bdf-a58e-8786e063c2fe.jpg" width="600">

Here are some other examples:
<img src="https://user-images.githubusercontent.com/47047889/161399096-cba4d8cc-1612-4781-b4f6-1ab3aae131f5.jpg" width="600">

<img src="https://user-images.githubusercontent.com/47047889/161399098-1bdde2b5-82bc-4692-943a-4cacfaf1c654.jpg" width="600">

<img src="https://user-images.githubusercontent.com/47047889/161399100-25badd05-2277-4a8b-b22c-faa6e34546f7.jpg" width="600">

As you can see, it is not perfect, but there are some similarity in terms of semantics.

Here is the code I used, basically, there are two differences: I center-cropped the image because this is how I deal with the training data, the hyper-parameter should be 0.25 instead of 0.75.

```
with torch.no_grad():
    preprocess = T.Compose([
                T.Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711)),
            ])  
    device = 'cuda:0'  # please use GPU, do not use CPU
    path = '/content/COCO2014_Language-free_Gaussian.pkl'  # pre-trained model

    # load pre-trained models
    generator = Generator(device=device, path=path)
    clip_model, _ = clip.load("ViT-B/32", device=device)
    clip_model = clip_model.eval()

    # putting image input for the generator
    real_img = Image.open(img_path).convert('RGB')
    img_array = T.ToTensor()(real_img).to(device)
    img_array = T.CenterCrop(size=min(img_array.shape[-1], img_array.shape[-2]))(img_array) # center crop the image
    reshaped_img = F.interpolate(preprocess(img_array).unsqueeze(0), (224, 224), mode='bicubic', align_corners=False) # resize to 224*224


    img_fts = clip_model.encode_image(reshaped_img)
    img_fts = img_fts/img_fts.norm(dim=-1, keepdim=True)
    img_fts = img_fts.repeat((num_images_to_generate, 1))

    random_noise = torch.randn(img_fts.shape).to(img_fts.device)
    random_noise = random_noise/random_noise.norm(dim=-1,keepdim=True)
    hyper = 0.25 # it should be 0.25 here, not 0.75
    img_fts = img_fts * hyper + random_noise * (1-hyper)
    img_fts = img_fts/img_fts.norm(dim=-1,keepdim=True)

    z = torch.randn((num_images_to_generate, 512)).to(device)
    c = torch.randn((num_images_to_generate, 1)).to(device)  # label is actually not used
    img, _ = generator.generate(z=z, c=c, fts=img_fts)

    original_img = (-1.+2.*F.interpolate(img_array.unsqueeze(0), (256, 256), mode='bicubic', align_corners=False)).repeat((num_images_to_generate, 1, 1, 1))
    to_show_img = generator.tensor_to_img(torch.cat([original_img, img], dim=-2))
    to_show_img.save('./generated.jpg')
```

**Comment by sbkim052** [comment]
Thanks. Now it is much better:)
One last question is about the random noise.
Why is the hyperparameter, which is multiplied to the random noise is bigger than the image feature?
What I thought was that the hyperparameter multiplied to image feature should be larger to preserve the semantic meaning of the input image.

**Comment by drboog** [comment]
This is hyper-parameter actually depends on the pre-trained CLIP model.

Let (X, T) be a ground-truth image-text pair, (x, t) be the corresponding features.

Although we may think that by contrastive learning, x and t should have very high cosine similarity: we expect the similarity to be close to 1. If you actually test a ground-truth pair from COCO, you will find that the (x, t) cosine similarity is around 0.3~0.4. 

As a result, given an image feature, we should perturb it so that the resulting similarity is around 0.3~0.4. You can quickly test that we can get the expected results with hyper-parameter in the range of 0.7~0.8.

You can see the the performance of language-free method is heavily influenced by the pre-trained model. I believe we can get a much better language-free generative model if we have a better pre-trained model, by which I mean the similarity of arbitrary paired features is close to 1. With the better pre-trained model, we don't need to use a large hyper-parameter. The fake text feature we generate is "much better", because large noise perturbation means larger probability to get a fake text feature that is out-of-distribution.

**Comment by sbkim052** [comment]
Thanks, @drboog:) I am now all clear.
I will close the issue.
